Unconditionally lower std::string's alignment requirement from 16 to 8.

EricWF opened the pull request:
Unconditionally change std::string's alignment to 8.

 This change saves memory by providing the allocator more freedom to allocate the most
 efficient size class by dropping the alignment requirements for std::string's
 pointer from 16 to 8. This changes the output of std::string::max_size, which makes it ABI breaking.

That said, the discussion concluded that we don't care about this ABI break. and would like this change enabled universally. 

The ABI break isn't one of layout or "class size", but rather the value of "max_size()" changes, which in turn changes whether `std::bad_alloc` or `std::length_error` is thrown for large allocations.

This change is the child of PR #68807, which enabled the change behind an ABI flag.

artagnon commented on libcxx/test/libcxx/strings/basic.string/string.capacity/max_size.pass.cpp:
```diff
@@ -18,7 +18,8 @@
 #include "test_macros.h"
 
 // alignment of the string heap buffer is hardcoded to 16
```
Please update this comment.

ldionne commented:
@EricWF Let's make this change!

EricWF committed:
Unconditionally change std::string's alignment to 8.

    Save memory by providing the allocator more freedom to allocate the most
    efficient size class by dropping the alignment requirements for std::string's
    pointer from 16 to 8. This changes the output of std::string::max_size,
    which makes it ABI breaking.

    That said, the discussion concluded that we don't care about this ABI
    break and would like this change enabled universally.

    The ABI break isn't one of layout or "class size", but rather the value of
    "max_size()" changes, which in turn changes whether std::bad_alloc or
    std::length_error is thrown for large allocations.

    This change is the child of PR #68807, which enabled the change behind an ABI
    flag.

ldionne reviewed (approved):
Nice, thanks!

ldionne commented:
It looks like this is breaking the backdeployment tests, which check the exact alignment. It should be sufficient to XFAIL them for older macOS deployment targets (anything before and including macOS 14.x which is the latest released version).

mstorsjo commented:
I see an ABI break being mentioned in the commit message, as being noted and agreed to be acceptable. I'm interested to hear about all the aspects of ABI break that was considered and deemed acceptable. In particular, won't this also break the layout of user's data structures? If I have e.g. `struct { char c; std::string str; }`, this struct will now have a different layout after this change, effectively breaking the ABI of potentially any struct/class that embeds a string?

ldionne commented on libcxx/include/__config:
```diff
@@ -174,11 +174,6 @@
 // The implementation moved to the header, but we still export the symbols from
 // the dylib for backwards compatibility.
 #    define _LIBCPP_ABI_DO_NOT_EXPORT_TO_CHARS_BASE_10
-// Save memory by providing the allocator more freedom to allocate the most
```
We need to adjust the release note in LLVM 18 to say that this isn't only done in the unstable ABI anymore.

EricWF replied:
Do you want this in the LLVM 18 release?

ldionne replied:
Yeah, I think I would put it in LLVM 18. That way we'd never ship the ABI macro and we'd avoid potential user churn.

EricWF replied:
Done.

ldionne commented on libcxx/include/__config:
```diff
@@ -174,11 +174,6 @@
 // The implementation moved to the header, but we still export the symbols from
```
Not attached to this line: the commit message should be self explanatory -- it shouldn't only contain a reference to the original discussion.

EricWF replied:
You mean the first comment on github or were you saying the actual message attached to the first commit wasn't sufficient?

I changed the first github comment to match the commit message with a little more info.

ldionne replied:
I meant the PR description on Github, since that is what Github uses to generate the commit message when we "squash and merge". Surprisingly, it seems that the actual commit messages are completely disregarded for that purpose.

EricWF replied:
Resolving. Please re-open if i misunderstood.

EricWF replied:
Urg, well thats gross and bad design on the part of github. 

I'll remember that going forward.

ldionne reviewed (commented):
@mstorsjo 
> I see an ABI break being mentioned in the commit message, as being noted and agreed to be acceptable. I'm interested to hear about all the aspects of ABI break that was considered and deemed acceptable.

This changes the value returned by `std::basic_string::max_size()`, which could in theory be relied upon, but in practice no sane code would ever rely on that for anything serious. In fact, the output of that function was wrong for 10+ years and we changed it a couple of releases ago. IMO changing the value returned by this function is as much of an ABI break as changing the return value of any other function to e.g. fix a bug -- it's really not much of a break.

> In particular, won't this also break the layout of user's data structures? If I have e.g. `struct { char c; std::string str; }`, this struct will now have a different layout after this change, effectively breaking the ABI of potentially any struct/class that embeds a string?

No, this only affects the alignment of the data allocated by the string in long mode. This doesn't change the layout of the `std::string` object itself.

If someone can spot that this is an ABI break in other ways, please let us know, but after thinking about it for a bit I can't think of anything.

mstorsjo commented:
> @mstorsjo
> 
> > I see an ABI break being mentioned in the commit message, as being noted and agreed to be acceptable. I'm interested to hear about all the aspects of ABI break that was considered and deemed acceptable.
> 
> This changes the value returned by `std::basic_string::max_size()`, which could in theory be relied upon, but in practice no sane code would ever rely on that for anything serious. In fact, the output of that function was wrong for 10+ years and we changed it a couple of releases ago. IMO changing the value returned by this function is as much of an ABI break as changing the return value of any other function to e.g. fix a bug -- it's really not much of a break.
> 
> > In particular, won't this also break the layout of user's data structures? If I have e.g. `struct { char c; std::string str; }`, this struct will now have a different layout after this change, effectively breaking the ABI of potentially any struct/class that embeds a string?
> 
> No, this only affects the alignment of the data allocated by the string in long mode. This doesn't change the layout of the `std::string` object itself.

Oh, ok - thanks, that clears my concern! Yes this sounds perfectly safe to me then.
